dtcli/validate_schema.py

```diff
@@ -0,0 +1,92 @@
+import json
+import pathlib
+from pathlib import Path
+import sys
+import yaml
+from typing import List, Union
+import enum
+
+
+YamlAST = Union[yaml.ScalarNode, yaml.SequenceNode, yaml.MappingNode]
+import jsonschema
+
+class AcquisitionMethod(enum.Enum):
+    Json = enum.auto()
+    Yaml = enum.auto()
+
+
+def backtrack_yaml_location(path: List[Union[int, str]], ast: YamlAST) -> yaml.error.Mark:
+    if not path:
+        return ast.start_mark 
+
+    chunk, *rest = path
+
+    if isinstance(chunk, str):
+        key_value_pairs = ast.value 
+        for k,v in key_value_pairs:
+            assert isinstance(k, yaml.ScalarNode), "keys are scalar nodes"
+            k = k.value
+            if k == chunk:
+                return backtrack_yaml_location(rest, v)
+    elif isinstance(chunk, int):
+        v = ast.value[chunk]
+        return backtrack_yaml_location(rest, v)
+    else:
+        assert 0, "unreachable switch branch"
+
+
+
+def validate_schema(instance_file: Path, schema_entrypoint: Path):
+    subschemas_path = pathlib.Path(schema_entrypoint).parent
+
+    try:
+        with open(instance_file, "r") as f:
+            instance, method = json.load(f), AcquisitionMethod.Json
+    except json.decoder.JSONDecodeError:
+        with open(instance_file, "r") as yaml_in:
+            instance, method = yaml.safe_load(yaml_in), AcquisitionMethod.Yaml
+
+    assert isinstance(method, AcquisitionMethod)
+
+    with open(schema_entrypoint) as f:
+        s = json.load(f)
+
+    schema_store = {}
+    for f in [f for f in pathlib.Path(subschemas_path).iterdir() if f.is_file()]:
+        try:
+            with open(f) as _f:
+                sub = json.load(_f)
+        except json.decoder.JSONDecodeError:
+            print(f"skipping subschema {f}, malformed json", file=sys.stderr)
+        else:
+            sub_id = sub["$id"]
+            schema_store[sub_id] = sub
+
+    resolver = jsonschema.validators.RefResolver.from_schema(s, store=schema_store)
+
+    _v = jsonschema.validators.validator_for(s)
+    _v.check_schema(s)
+    v = _v(s, resolver=resolver)
+
+    with open(instance_file, "r") as f:
+        file = yaml.compose(f)
+
+    errors = []
+    for i, error in enumerate(v.iter_errors(instance)):
+        # TODO: add line info - might need to differenciate between json and yaml source, currently we'll focus on yaml and provide a null response on json
+        err_loc = backtrack_yaml_location(error.absolute_path, file)
+
+        # TODO: add typo finder for some cases - like enum mismatch
+        errors.append({
+        "error_num" : i,
```

Comment:
Why export error number? That is only relevant for presentation reason and should be done at the UI layer